### PR TITLE
Upgradeable component

### PR DESCRIPTION
**Pull request by notV4l**

### CODE REVIEW

File: crates/dojo-core/src/components/upgradeable.cairo [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+    impl Upgradable<
+        TContractState, +HasComponent<TContractState>, +IWorldProvider<TContractState>
+    > of super::IUpgradeable<ComponentState<TContractState>> {
+        fn upgrade(ref self: ComponentState<TContractState>, new_class_hash: ClassHash) {
+            assert(self.get_contract().world().contract_address.is_non_zero(), Errors::INVALID_WORLD_ADDRESS);
+            assert(get_caller_address().is_non_zero(), Errors::INVALID_CALLER);
+            assert(get_caller_address() == self.get_contract().world().contract_address, Errors::INVALID_CALLER);
+            assert(new_class_hash.is_non_zero(), Errors::INVALID_CLASS);
+           
+            replace_class_syscall(new_class_hash).unwrap();
```
[the remark this answers is not in the document]
Comment (notV4l) [reply crates/dojo-core/src/components/upgradeable.cairo:52]: yes ser

File: crates/dojo-lang/src/contract.rs [hunk 2]
```diff
@@ -48,29 +48,38 @@ impl DojoContract {
...
 
+                    component!(path: upgradeable_component, storage: upgradeable, event: \
+                 UpgradeableEvent);
+
+                    
                     #[storage]
                     struct Storage {
                         world_dispatcher: IWorldDispatcher,
+                        #[substorage(v0)]
+                        upgradeable: upgradeable_component::Storage,
```
Comment (tarrencev) [inline crates/dojo-lang/src/contract.rs:63]: better to fully qualify here and avoid import conflicts

```suggestion
                        upgradeable: dojo::components::upgradeable::upgradeable::Storage,
```

File: crates/dojo-lang/src/contract.rs [hunk 3]
```diff
@@ -101,58 +110,29 @@ impl DojoContract {
...
+        let variants = variants.join(", ");
 
         rewrite_nodes.push(RewriteNode::interpolate_patched(
             "
-                #[external(v0)]
-                fn execute(self: @ContractState, $params$) $ret_clause$ $body$
+            #[event]
+            #[derive(Drop, starknet::Event)]
+            enum Event {
+                UpgradeableEvent: upgradeable_component::Event,
```
Comment (tarrencev) [inline crates/dojo-lang/src/contract.rs:130]: Here too

**Review by tarrencev — APPROVED** [review]

**Comment by tarrencev** [comment]
can we move the erc20 / erc721 stuff to origami now

**Comment by notV4l** [comment]
> can we move the erc20 / erc721 stuff to origami now

yes dojo-erc aswell as dojo-defi & dojo-physics has moved to origami and should be deleted from dojo

**Comment by tarrencev** [comment]
do you want to do that before we land this?

**Comment by notV4l** [comment]
> do you want to do that before we land this?

yes lets do it

**Review by tarrencev — APPROVED** [review]

**Pull request merged**